### PR TITLE
feat(TF>=1.1)!: Configure ASM management mode

### PR DESCRIPTION
**Pull request by ferrarimarco**
This PR implements the ability of configuring ASM management mode.

Without this, users have to manually configure the feature membership.

### CODE REVIEW

**Comment by ferrarimarco** [comment]
Ah! But we configure the CPR already, so this may not actually be needed...

**Comment by ferrarimarco** [comment]
Also, it looks like that the `module.cpr` and `kubernetes_config_map.asm_options` aren't necessary anymore because the fleet API takes care of that, and actually conflict with the Managed ASM control plane installation. Worth removing in this PR, or a separate one?

**Comment by ferrarimarco** [comment]
PS: the build failure doesn't seem to be related to this PR.

**Comment by apeabody** [comment]
Thanks for the pull request @ferrarimarco 

From the linter
```
Checking submodule's files generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/asm/README.md /tmp/tmp.eM1y9hxVzy/workspace/modules/asm/README.md
62c62
< | enable\_vpc\_sc | Determines whether to enable VPC-SC for this ASM installation. For more information read <https://cloud.google.com/service-mesh/docs/managed/vpc-sc> | `bool` | `false` | no |
---
> | enable\_vpc\_sc | Determines whether to enable VPC-SC for this ASM installation. For more information read https://cloud.google.com/service-mesh/docs/managed/vpc-sc | `bool` | `false` | no |
65c65
< | mesh\_management | ASM management mode. For more information, see the [gke_hub_feature_membership resource documentation](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/gke_hub_feature_membership#nested_mesh) | `string` | `""` | no |
---
> | mesh\_management | ASM Management mode. | `string` | `""` | no |
Error: submodule's files generation has not been run, please run the
'make build' command and commit changes

Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=autogen' '--exclude=*.tfvars' '--exclude=*metadata.yaml' /workspace/modules/asm/README.md /tmp/tmp.3rgofwWMtd/generate_docs/workspace/modules/asm/README.md
62c62
< | enable\_vpc\_sc | Determines whether to enable VPC-SC for this ASM installation. For more information read <https://cloud.google.com/service-mesh/docs/managed/vpc-sc> | `bool` | `false` | no |
---
> | enable\_vpc\_sc | Determines whether to enable VPC-SC for this ASM installation. For more information read https://cloud.google.com/service-mesh/docs/managed/vpc-sc | `bool` | `false` | no |
65c65
< | mesh\_management | ASM management mode. For more information, see the [gke_hub_feature_membership resource documentation](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/gke_hub_feature_membership#nested_mesh) | `string` | `""` | no |
---
> | mesh\_management | ASM Management mode. | `string` | `""` | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
```

**Comment by ferrarimarco** [comment]
@apeabody thanks for the update. I'll look into that.

**Comment by ferrarimarco** [comment]
@apeabody hopefully fixed!

**Comment by apeabody** [comment]
> @apeabody hopefully fixed!

Thanks @ferrarimarco - With the addition of count to the module, you will also need to add the index for the output value:

```
│ Error: Unsupported attribute
│ 
│   on ../../../modules/asm/outputs.tf line 23, in output "wait":
│   23:   value       = module.cpr.wait
│     ├────────────────
│     │ module.cpr is a list of object
│ 
│ Can't access attributes on a list of objects. Did you mean to access
│ attribute "wait" for a specific element of the list, or across all elements
│ of the list?
```

**Comment by ferrarimarco** [comment]
@apeabody thanks for your review.

I've modified the `wait` output variable initialization to fix the issue reported in https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/1702#issuecomment-1741107235.

Also, the module now accordingly sets its value considering both `MANAGEMENT_AUTOMATIC` and `MANAGEMENT_MANUAL` (or unset management mode).

File: modules/asm/variables.tf [hunk 1]
```diff
@@ -50,6 +50,20 @@ variable "channel" {
   default = ""
 }
 
+variable "mesh_management" {
+  default     = ""
```
Comment (apeabody) [inline modules/asm/variables.tf:54]: Might be nice to also support `null`.

Would need to bump the min TF version to `1.1`: https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/modules/asm/versions.tf#L19

Comment (ferrarimarco) [reply modules/asm/variables.tf:54]: Done!

**Comment by ferrarimarco** [comment]
Build failures seem related to quota issues.

**Comment by ferrarimarco** [comment]
This failure seem to be related to an unsupported GKE version.

**Comment by apeabody** [comment]
> Also, it looks like that the `module.cpr` and `kubernetes_config_map.asm_options` aren't necessary anymore because the fleet API takes care of that, and actually conflict with the Managed ASM control plane installation. Worth removing in this PR, or a separate one?

Hi @ferrarimarco - Is there a recommend migration path?  While reviewing this PR its clear that supporting both adds extra complexity.  Ideally we would want to minimze wrapping imperative commands like in [`module.cpr`](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/modules/asm/main.tf#L58).

**Comment by ferrarimarco** [comment]
Currently, Anthos Service Mesh supports the following installation methods:

- Managed ASM: managed control plane that runs outside the cluster.
  - [using Fleet API](https://cloud.google.com/service-mesh/docs/managed/provision-managed-anthos-service-mesh): fully automated, recommended by ASM docs.
  - [using asmcli](https://cloud.google.com/service-mesh/docs/managed/provision-managed-anthos-service-mesh-asmcli).
- In-cluster ASM: control plan runs inside the cluster. [Install using asmcli](https://cloud.google.com/service-mesh/docs/unified-install/install-anthos-service-mesh).

In all cases, there's no explicit creation of `ControlPlaneRevision` resources. Either the Fleet API does it, or `asmcli` takes care of creating it.

My opinion is that we should:

- Remove the `module.cpr` resource because the Fleet API and `asmcli` take care of creating it.
- Remove the `kubernetes_namespace.system` resource because the Fleet API and `asmcli` take care of creating it.
- Clarify that we install the "managed ASM control plane" version using the Fleet API.
- Refactor optional features according to [Enable optional features on managed ASM](https://cloud.google.com/service-mesh/docs/managed/enable-managed-anthos-service-mesh-optional-features). This probably requires dropping unsupported features, such as `var.enable_cni`.

Installation methods that involve using `asmcli` will likely result in having to wrap calls to `asmcli` using `null_resource`s, which are hard to maintain and to implement with the right triggers. That's why I suggest sticking with provisioning and configuring managed ASM using the Fleet API only.

I'm happy to talk about this over Meet if you prefer :)

/cc @jbrook

**Comment by apeabody** [comment]
> > Currently, Anthos Service Mesh supports the following installation methods:
> > 
> > * Managed ASM: managed control plane that runs outside the cluster.
> >   
> >   * [using Fleet API](https://cloud.google.com/service-mesh/docs/managed/provision-managed-anthos-service-mesh): fully automated, recommended by ASM docs.
> >   * [using asmcli](https://cloud.google.com/service-mesh/docs/managed/provision-managed-anthos-service-mesh-asmcli).
> > * In-cluster ASM: control plan runs inside the cluster. [Install using asmcli](https://cloud.google.com/service-mesh/docs/unified-install/install-anthos-service-mesh).
> > 
> > In all cases, there's no explicit creation of `ControlPlaneRevision` resources. Either the Fleet API does it, or `asmcli` takes care of creating it.
> > My opinion is that we should:
> > 
> > * Remove the `module.cpr` resource because the Fleet API and `asmcli` take care of creating it.
> > * Remove the `kubernetes_namespace.system` resource because the Fleet API and `asmcli` take care of creating it.
> > * Clarify that we install the "managed ASM control plane" version using the Fleet API.
> > * Refactor optional features according to [Enable optional features on managed ASM](https://cloud.google.com/service-mesh/docs/managed/enable-managed-anthos-service-mesh-optional-features). This probably requires dropping unsupported features, such as `var.enable_cni`.
> > 
> > Installation methods that involve using `asmcli` will likely result in having to wrap calls to `asmcli` using `null_resource`s, which are hard to maintain and to implement with the right triggers. That's why I suggest sticking with provisioning and configuring managed ASM using the Fleet API only.
> > I'm happy to talk about this over Meet if you prefer :)
> > /cc @jbrook
> 
> Thanks @ferrarimarco - Given the scale of the change, let me ping you directly.

Hi @ferrarimarco - Did we want to move forward with this change for now?  Cheers!

**Comment by ferrarimarco** [comment]
Hi Andrew! I'll align internally and get back you ASAP. Thanks for following up!

File: modules/asm/main.tf [hunk 2]
```diff
@@ -34,14 +34,16 @@ data "google_container_cluster" "asm" {
...
-  count = var.create_system_namespace ? 1 : 0
+  count = var.create_system_namespace && var.mesh_management != "MANAGEMENT_AUTOMATIC" ? 1 : 0
 
   metadata {
     name = "istio-system"
   }
 }
 
 resource "kubernetes_config_map" "asm_options" {
+  count = var.mesh_management != "MANAGEMENT_AUTOMATIC" ? 1 : 0
```
Comment (apeabody) [inline modules/asm/main.tf:45]: With TF 1.1 (perhaps earlier as well), when adding `count` to an existing resource, Terraform will automatically move state to index 0, so these should be fine.

File: modules/asm/main.tf [hunk 3]
```diff
@@ -69,3 +73,42 @@ module "cpr" {
...
+module "kubectl_asm_wait_for_controlplanerevision_custom_resource_definition" {
+  count = var.mesh_management == "MANAGEMENT_AUTOMATIC" ? 1 : 0
+
+  source  = "terraform-google-modules/gcloud/google//modules/kubectl-wrapper"
+  version = "~> 3.1"
+
+  project_id              = var.project_id
+  cluster_name            = var.cluster_name
+  cluster_location        = var.cluster_location
+  kubectl_create_command  = "/bin/sh -c 'while ! kubectl wait crd/controlplanerevisions.mesh.cloud.google.com --for condition=established --timeout=60m --all-namespaces; do echo \"crd/controlplanerevisions.mesh.cloud.google.com not yet available, waiting...\"; sleep 5; done'"
```
[the remark this answers is not in the document]
Comment (apeabody) [reply modules/asm/main.tf:88]: As this is only for the new `MANAGEMENT_AUTOMATIC` mode, it won't change the behaivor for existing clusters.

**Review by apeabody — APPROVED** [review]
Thanks @ferrarimarco!

**Pull request merged**